### DISCUSSION TITLE
Add option to render maps using StellarMaps (if installed)

### DISCUSSION THREAD

**Discussion by oatmealproblem** [eliasdoehne/stellaris-dashboard#167] [Ideas]
My main side project recently has been StellarMaps https://github.com/MichaelMakesGames/stellarmaps, a Stellaris map renderer. I think it would be very cool if the dashboard could show maps produced by StellarMaps, if both are installed and running. The dashboard would get additional features (terra incognita, country emblems, colored borders, style customization etc), plus any new features added to StellarMaps.

This isn't possible now (I would need to add some features to StellarMaps), so this is just to start the conversation. Main thing to figure out is how the dashboard would get the images from StellarMaps. I see 2 basic approaches:

- stellaris-dashboard requests a map render, either by running a StellarMaps executable, or request to a StellarMaps web server running on localhost
- StellarMaps simply exports a PNG or SVG to a well-defined location for each new save file detected, and the dashboard checks that location
  - I'm already planning on adding most of the features required for this, as part of my plan to add time-lapse functionality to StellarMaps

By the way, here's a few pictures of maps generated with StellarMaps:
![lotor_cropped](https://github.com/eliasdoehne/stellaris-dashboard/assets/5983927/fd80c841-1d89-44b6-9f3c-dbab489e3e5e)
![map](https://github.com/eliasdoehne/stellaris-dashboard/assets/5983927/563da995-deb3-4aa4-b79e-f6ffd047838e)
![Screenshot from 2023-09-11 16-57-05](https://github.com/eliasdoehne/stellaris-dashboard/assets/5983927/8eed0b09-d5f0-4b2b-a480-d5fae4819d06)

**Comment by eliasdoehne** [comment]
Wow, that looks really nice! I'll try it out this weekend. If there is a good way to integrate it, let's try it!

For reference, the galaxy map in the UI is built with the dash framework [here](https://github.com/eliasdoehne/stellaris-dashboard/blob/00b4f03b19c6bb51cf6cfae907a8edbf1c875c81/stellarisdashboard/dashboard_app/graph_ledger.py#L560) and there is an alternative  [matplotlib implementation](https://github.com/eliasdoehne/stellaris-dashboard/blob/00b4f03b19c6bb51cf6cfae907a8edbf1c875c81/stellarisdashboard/dashboard_app/timelapse_exporter.py#L127) that is used for exporting timelapses.

**Comment by oatmealproblem** [comment]
Thanks!

The more I think about it, the more I think StellarMaps should just be running in the background, and Stellaris Dashboard should just load the files it produces. StellarMaps needs data from the saves that Stellaris Dashboard does not (emblems, terra incognita, and more).

I think SVG format will probably be best. That should be smaller than the PNGs, it's more efficient for StellarMaps to produce, and it will work much better than PNGs if we want it to be interactive. A simple POC implementation could just serve up a the most recent StellarMaps export. Then later we can add timeline controls, panning/zooming, and system tooltips.

Fow now, StellarMaps does not have the ability to run in the background, but I'm going to add that away, planned for the release after next. Can't do much on this until then.

**Comment by eliasdoehne** [comment]
Sounds good,I was also considering to backup the saves parsed by the dashboard to some location to allow reprocessing. Maybe it would also be an option to pass the save to your application on demand for time-travel or time lapses?

Just fyi, I'm gone for a few days but I should be able to take some time this weekend

**Comment by oatmealproblem** [comment]
That could work too.

Depending on autosave frequently, that could be lot of data you're backing up. Save files can be many megabytes, especially on a larger map or later in the game, so that could mean gigabytes worth of back ups. For StellarMaps I'm planning on creating save backups, but only the subset of data StellarMaps so it doesn't use too much disk space

> Just fyi, I'm gone for a few days but I should be able to take some time this weekend

No worries and no rush, we're all just in this for fun

**Comment by oatmealproblem** [comment]
Unrelated to this, I plan on creating a feature-limited AWS Lambda version of StellarMaps. That *might* also be the best way to integrate StellarMaps with Stellaris Dashboard. Then, the user wouldn't even need to StellarMaps installed locally. I'll post here again after I've actually implemented that

**Comment by oatmealproblem** [comment]
StellarMaps is undergoing a lot architectural changes right now. I'm going to convert this into a discussion, since there's no clear path forward (not saying it's not possible, just best to wait for the next iteration of StellarMaps to stabilize)
